kiwi/storage/subformat/vagrant_libvirt.py

```diff
@@ -83,71 +44,24 @@ def create_image_format(self):
                 box_img
             ]
         )
+        return [box_img]
 
-        metadata_json = os.sep.join([self.temp_image_dir, 'metadata.json'])
-        with open(metadata_json, 'w') as meta:
-            meta.write(self._create_box_metadata())
-
-        vagrantfile = os.sep.join([self.temp_image_dir, 'Vagrantfile'])
-        with open(vagrantfile, 'w') as vagrant:
-            vagrant.write(self._create_box_vagrantconfig())
-
-        Command.run(
-            [
-                'tar', '-C', self.temp_image_dir,
-                '-czf', self.get_target_file_path_for_format(self.image_format),
-                os.path.basename(box_img),
-                os.path.basename(metadata_json),
-                os.path.basename(vagrantfile)
-            ]
-        )
-
-    def store_to_result(self, result):
+    def get_additional_metadata(self):
+        # () -> Dict[str, str]
         """
-        Store result file of the vagrant format conversion into the
-        provided result instance. In this case compression is unwanted
-        because the box is already created as a compressed tarball
-
-        :param object result: Instance of Result
+        Returns a dictionary containing the virtual image format and the size
+        of the image.
         """
-        result.add(
-            key='disk_format_image',
-            filename=self.get_target_file_path_for_format(
-                self.image_format
-            ),
-            use_for_bundle=True,
-            compress=False,
-            shasum=True
-        )
-
-    def _create_box_metadata(self):
-        metadata = {
-            'provider': 'libvirt',
+        return {
             'format': 'qcow2',
             'virtual_size': format(self.vagrantconfig.get_virtualsize() or 42)
         }
-        return json.dumps(
-            metadata, sort_keys=True, indent=2, separators=(',', ': ')
-        )
-
-    def _create_box_vagrantconfig(self):
-        vagrantconfig = dedent('''
-            Vagrant::Config.run do |config|
-              config.vm.base_mac = "{mac_address}"
-            end
-            include_vagrantfile = File.expand_path(
-              "../include/_Vagrantfile", __FILE__
-            )
-            load include_vagrantfile if File.exist?(include_vagrantfile)
-        ''').strip()
-        return vagrantconfig.format(
-            mac_address=self._random_mac()
-        )
 
-    def _random_mac(self):
-        return '%02x%02x%02x%02x%02x%02x'.upper() % (
-            0x00, 0x16, 0x3e,
-            random.randrange(0, 0x7e),
-            random.randrange(0, 0xff),
-            random.randrange(0, 0xff)
-        )
+    def additional_vagrant_settings(self):
+        # () -> str:
+        """
+        Returns settings for the libvirt provider telling vagrant to use kvm.
+        """
+        return """config.vm.provider :libvirt do |libvirt|
+    libvirt.driver = "kvm"
+  end"""
```

Comment:
Could you use the dedent module for returning strings like this:

```python
from textwrap import dedent

libvirt_driver_setup = dedent('''
    config.vm.provider :libvirt do |libvirt|
        libvirt.driver = "kvm"
    end
''')

return libvirt_driver_setup
```

This is imho nicer to read and explicit in what it returns

Reply:
I'll give it a try, but I think I didn't use it because it broke the indentation of the resulting `Vagrantfile` that ends up in the box.

Reply:
The dedent module allows to provide the indentation as you need it. If you strip() and add a linesep at the end it should be identical to the version you had in """ quotes

Reply:
I have changed the way `_create_box_vagrantconfig` pastes the customized vagrant config and the result looks better now, albeit requiring another call to `indent`.

Reply:
from textwrap import dedent

that exists in py2 and py3, kiwi already uses that module for years :)

Reply:
Still like the version from above more. Sorry but

```
 "  " + "  ".join(
            dedent("""
```

looks strange to me. Also python uses single quotes for text that is taken one by one.

```python
dedent('''
    config.vm.provider :libvirt do |libvirt|
        libvirt.driver = "kvm"
    end
''').strip()
```

produces the same formatting result and avoids and leading spaces. Any reason why that does not work for you ?

Reply:
Also saw you are using splitlines() now which turns the string including espace sequences into a list. That invalidates the doc string because the method now returns a list and not a string anymore. Also why is this better than before ? It's starting to confuse me.

Reply:
The concept of additional_vagrant_settings() is ok for me as it just added raw data to the vagrantfile. Maybe the method should be better named additional_vagrant_raw_config() or similar. If you wish to add more structure I suggest to pass a dictionary with information, but in case of the Vagrantfile which is actually ruby code the idea to pass raw data is probably the more lightweight solution. So I'm ok with the string but the list I don't understand

Reply:
Yes, I actually need the two leading spaces, otherwise it'll mess up the `Vagrantfile`.

Since it is quite confusing having to do that in child classes, I have moved the indentation code to `DiskFormatVagrantBase._create_box_vagrantconfig`, so that this is done once for all children.

Reply:
ok, thanks for the changes. I was reading the code again and I think we should do the vagrantconfig handling a bit different. My idea is to use a simple template. I hope you don't mind but that spacing and such could bite us in the future and is not very flexible. So here is my idea

```python
import os
  
from textwrap import dedent
from string import Template

class VagrantConfigTemplate(object):
    """
    **vagrant config template**
    """
    def __init__(self):
        self.indent = '  ';

        self.header = dedent('''
            Vagrant.configure("2") do |config|
        ''').strip() + os.linesep

        self.mac = self.indent + dedent('''
            config.vm.base_mac = ${mac_address}
        ''').strip() + os.linesep

        self.end = dedent('''
            end
        ''').strip()

    def get_template(self, custom_settings=None):
        template_data = self.header
        template_data += self.mac
        if custom_settings:
            template_data += self.indent
            template_data += self.indent.join(
                custom_settings.splitlines(True)
            )
            template_data += os.linesep
        template_data += self.end
        return Template(template_data)

def get_additional_vagrant_config_settings():
    return dedent('''
        config.vm.provider :libvirt do |libvirt|
          libvirt.driver = "kvm"
        end
    ''').strip()

def _create_box_vagrantconfig():
    template = VagrantConfigTemplate()
    vagrant_config = template.get_template(
        custom_settings=get_additional_vagrant_config_settings()
    )
    return vagrant_config.substitute(
        {'mac_address': 'foo'}
    )

print(_create_box_vagrantconfig())
```

What do you think ? It will connect all indention and layouting into a template class and also allows for better testing.

Thanks for your patience

Reply:
I'll give your idea a try and code something up so that we find out its pain points. I agree that just pasting stuff into the `Vagrantfile` (as I have done it) is pretty ugly and could result in bad things happening.

Reply:
great, thanks :+1:

Reply:
Done. I have refactored the code a little bit and converted it to a single function. Since I added a doctest to this function I tried to enable them for pytest, unfortunately it does not seem to collect them properly.